Move random referral's app to rails

beeLexington opened the pull request:
Moved the app from a sinatra app over into a rails app.
The app has been stripped back a little. The form has been removed and replaced with a simple list within the index page.

Next steps are to add in styling

beeLexington committed:
Moved sinatra app into subdirectory

beeLexington committed:
Set up blank rails project

beeLexington committed:
Render a referral code on a page

beeLexington committed:
Set up rspec

beeLexington committed:
Add dependendies needed to use GoogleDrive::Session

beeLexington committed:
Stop hard coding referral info in model

beeLexington committed:
Import referrals from Google Sheets

We mock Referral.worksheet to return example data, covering potential
duplication.

We filter out referrals which haven't opted in.

beeLexington committed:
Show a random referral code

beeLexington committed:
Support showing a specific referral

We add a /random route which redirects to the show route after picking
a random referral. We want referrals to be bookmarkable and shareable
so we need a consistent link to a specific referral.

To do this we need a unique id for a referral. We use a hash of the
service provider combined with the code for the id. We use a hash
because codes are often urls and create ugly paths when parameterised.

beeLexington committed:
Add service provider to model

beeLexington committed:
Expose machine readable referral service provider

beeLexington committed:
Support selecting a random referral by service provider

We want to be able to show a random referral for a specific service
provider.

beeLexington committed:
Allow selecting a referral by service provider

This adds an index page listing all of the service providers. Clicking
one service provider sends you to a referral for that service provider.

beeLexington committed:
Expose the name of the person for a referral

beeLexington committed:
Display the name of the referral's person

Floppy commented on random-referral/app/models/referral.rb:
```diff
@@ -23,6 +23,7 @@ def self.all
       }.map { |row|
         # columns are zero-indexed
         Referral.new(
+          id: row[1].gsub(/[\W]+/, "_").downcase + "_" + row[0].gsub(/[\W]+/, "_").downcase,
```
This is untested; we probably need to extend the tests for `.all` to check that the actual data in the Referral objects created is correct, and need to update the mock worksheet that comes back.

Floppy replied:
And the same for service provider too, needs adding into the `.all` tests

Floppy replied:
mock worksheet was updated in a later commit, so that's fine, but there is still a test missing for the actual content parsing; having that test earlier would have meant the worksheet would have needed to be updated at the same time, so would all be a bit more internally consistent.

Floppy commented on random-referral/app/models/referral.rb:
```diff
@@ -41,4 +41,8 @@ def initialize(id:, code:, service_provider:)
   end
+
+  def service_provider_id
+    @service_provider.gsub(/[\W]+/, "_").downcase
```
I suspect you could do this with https://apidock.com/rails/v5.2.3/ActiveSupport/Inflector/underscore

Floppy replied:
or perhaps https://apidock.com/rails/v5.2.3/ActiveSupport/Inflector/parameterize is right

Floppy replied:
parameterize might be the right one for `id`, to use in URLs. Probably doesn't matter which you use for this one?

Floppy commented on README.md:
```diff
@@ -1,25 +1,24 @@
-# Random Referral
+# README
```
move the old README back, this is the default Rails one :)

Floppy commented on app/models/referral.rb:
```diff
@@ -0,0 +1,58 @@
+  def self.worksheet
+    service_account_io = OpenStruct.new(read: "{
+      \"type\": \"service_account\",
+      \"project_id\": \"#{ENV["GOOGLE_DRIVE_PROJECT_ID"]}\",
```
Not essential but consider using `ENV#fetch` here for more predictable explosions if there are missing ENV variables.

Floppy commented on app/views/referrals/index_by_service_provider.html.erb:
```diff
@@ -0,0 +1,9 @@
+<ul>
+  <% @service_providers.map do |service_provider| %>
+    <li>
+      <a href=<%=random_referral_path(service_provider: service_provider[:id])%>><%= service_provider[:name] %></a>
```
more railsy to use `link_to` perhaps?

Floppy commented on spec/models/referral_spec.rb:
```diff
@@ -0,0 +1,91 @@
+
+      expect(referrals).not_to be_empty
+
+      referrals.each do |referral|
```
Might be more rubyish to say `referral.map{|x| x.is_a?(Referral)}.all?`, but perhaps also it's less readable. Not sure, but consider. https://ruby-doc.org/core-3.1.2/Enumerable.html#method-i-all-3F

Floppy reviewed (changes requested):
The important bits:

* Add some context (or combine with other commits) why ActiveModel::Model gets added to Referral - that wasn't clear what it was for
* Add missing test for worksheet column parsing in `.all`
* use `parameterize` when converting strings for use in URLs

Other things are mainly just suggestions or options. 👍🏻

beeLexington committed:
Replace sinatra app with rails app

beeLexington committed:
Install and run standardrb

To run standardrb `bundle exec standardrb --fix`

https://github.com/testdouble/standard

beeLexington committed:
Update .gitignore

To add artefacts from vscode IDE

beeLexington committed:
Generate referral id in initialize

We now have everything we need to generate the id within initialize.
This means all referral ids use the same structure regardless of where
they were created.

beeLexington commented:
> The important bits:
> 
>     * Add some context (or combine with other commits) why ActiveModel::Model gets added to Referral - that wasn't clear what it was for
> 
>     * Add missing test for worksheet column parsing in `.all`
> 
>     * use `parameterize` when converting strings for use in URLs
> 
> 
> Other things are mainly just suggestions or options. 👍🏻

I have addressed these points and believe I am ready for a re-review!

lozette commented on random-referral/Gemfile:
```diff
@@ -38,8 +38,9 @@ gem "bootsnap", require: false
-  gem 'rspec-rails', '~> 6.0.0.rc1'
+  gem "debug", platforms: %i[mri mingw x64_mingw]
+  gem "rspec-rails", "~> 6.0.0.rc1"
+  gem "standard"
```
Can you add a note to the README about running this locally

beeLexington replied:
Ah yes, I forgot to update the README, thanks for the reminder!

lozette commented on random-referral/app/controllers/referrals_controller.rb:
```diff
@@ -1,5 +1,12 @@
 class ReferralsController < ApplicationController
   def show
+    @referral = Referral.all.find do |referral|
+      referral.id == params[:id]
```
Do we need to gracefully handle a case where the referral ID is not found?

beeLexington replied:
Do you mean to render an error message if ID is not found?

lozette replied:
Yes, I'm not sure what will happen if the ID is not found - will it show a generic Rails not found error? Do we want it to show a custom error page? I don't think it's hugely important however.

lozette reviewed (approved):
Just a couple of comments, not blockers but it would be nice to address them before merge

beeLexington committed:
Update README to reflect the move to Rails

beeLexington committed:
Enable user friendly error handling

Floppy reviewed (approved):
Looks good to me! Nice work @beeLorna :)
